services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/geometria/GeometriatServiceImpl.kt

```diff
@@ -2,68 +2,72 @@ package fi.hel.haitaton.hanke.geometria
 import java.time.ZonedDateTime
 import mu.KotlinLogging
+import org.geojson.FeatureCollection
 import org.springframework.transaction.annotation.Transactional
 
 private val logger = KotlinLogging.logger {}
 
 open class GeometriatServiceImpl(private val hankeGeometriaDao: GeometriatDao) : GeometriatService {
 
     @Transactional
-    override fun saveGeometriat(geometriat: Geometriat): Geometriat {
+    override fun saveGeometriat(geometriat: HasFeatures, existingId: Int?): Geometriat? {
         GeometriatValidator.expectValid(geometriat)
 
-        val now = ZonedDateTime.now(TZ_UTC)
-        val oldGeometriat =
-            if (geometriat.id != null) hankeGeometriaDao.retrieveGeometriat(geometriat.id!!)
-            else null
+        val oldGeometriat = existingId?.let { hankeGeometriaDao.retrieveGeometriat(existingId) }
         val hasFeatures = geometriat.hasFeatures()
-        val username = currentUserId()
 
         return when {
             oldGeometriat == null && !hasFeatures ->
                 throw IllegalArgumentException("New Geometriat does not contain any Features")
-            !hasFeatures -> {
-                // DELETE
-                geometriat.id = oldGeometriat!!.id
-                geometriat.createdByUserId = oldGeometriat.createdByUserId
-                geometriat.createdAt = oldGeometriat.createdAt
-                geometriat.modifiedByUserId = username
-                geometriat.modifiedAt = now
-                geometriat.version = oldGeometriat.version!! + 1
-                hankeGeometriaDao.deleteGeometriat(geometriat)
-                logger.info { "Deleted geometries ${geometriat.id}" }
-                geometriat
-            }
-            oldGeometriat == null -> {
-                // CREATE
-                geometriat.createdByUserId = username
-                geometriat.createdAt = now
-                geometriat.modifiedByUserId = null
-                geometriat.modifiedAt = null
-                geometriat.version = 0
-                val created = hankeGeometriaDao.createGeometriat(geometriat)
-                logger.info { "Created new geometries ${created.id}" }
-                created
-            }
-            else -> {
-                // UPDATE
-                if (oldGeometriat.version == null) {
-                    error("There is an old Geometriat ${oldGeometriat.id} but it has no 'version'")
-                } else {
-                    oldGeometriat.version = oldGeometriat.version!! + 1
-                }
-                oldGeometriat.modifiedByUserId = username
-                oldGeometriat.modifiedAt = now
-                oldGeometriat.featureCollection = geometriat.featureCollection
-                hankeGeometriaDao.updateGeometriat(oldGeometriat)
-                logger.info { "Updated geometries ${oldGeometriat.id}" }
-                oldGeometriat
-            }
+            oldGeometriat != null && !hasFeatures -> deleteGeometriat(oldGeometriat)
```

Comment:
Renamed as suggested. I also moved the `hasFeatures()` and `resetFeatureProperties()` functions inside the HasFeatures interface, since they were declared right beside it anyway.